File: [0128][Hard][Longest_Consecutive_Sequence][UnionFind]/Longest_Consecutive_Sequence.java
// solution 1, union find, 
// leetcode time     cost : 1040 ms
// leetcode memory   cost : 42.2 MB 
// Time  Complexity: O(N)
// Space Complexity: O(N)
import java.util.HashMap;
import java.util.Map;
class Solution {
    public int find(int x,Map<Integer,Integer> map) {
        return map.containsKey(x)?find(x+1, map):x;
    }
    public int longestConsecutive(int[] nums) {
        Map<Integer,Integer> map = new HashMap<>();
        //make set
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        int max=0;
        //find
        for (int i:nums) {       
            max=Math.max(max,find(i+1 , map)-i);
        }
        return max;
    }
}

public class Longest_Consecutive_Sequence {
    public static void main(String args[]) {
        int nums[] = {100, 4, 200, 1, 3, 2}; // #expect is 4
        Solution Solution_obj = new Solution();
        int result = Solution_obj.longestConsecutive(nums);
        System.out.println("In java code,result is :" + result);
    }
}

File: [0128][Hard][Longest_Consecutive_Sequence][UnionFind]/Longest_Consecutive_Sequence.py
# solution 1, DP,UnionFind
# leetcode time     cost : 44 ms
# leetcode memory   cost : 14.8 MB 
# Time  Complexity: O(N)
# Space Complexity: O(N)
class Solution(object):
    def longestConsecutive(self, nums):
        hash_dict = dict()
        
        max_length = 0
        for num in nums:
            if num not in hash_dict:
                left = hash_dict.get(num - 1, 0)
                right = hash_dict.get(num + 1, 0)
                
                cur_length = 1 + left + right
                if cur_length > max_length:
                    max_length = cur_length
                
                hash_dict[num] = cur_length
                hash_dict[num - left] = cur_length
                hash_dict[num + right] = cur_length
        return max_length

def main():
    inputX,expectRes = [100, 4, 200, 1, 3, 2], 4
    obj = Solution()
    result = obj.longestConsecutive(inputX)
    try:
        assert result == expectRes
        print("passed, result is follow expect:",result)
    except AssertionError as aError:
        print('failed, result >> ', result,"<< is wrong, ","expect is : "+ expectRes)
    
if __name__ =='__main__':
    main()  